sabnzbd/newsunpack.py

```diff
@@ -1918,3 +1919,92 @@ def open(self, name: str) -> BinaryIO:
     def close(self):
         """Close file"""
         pass
+
+
+##############################################################################
+# .tar Functions
+##############################################################################
+
+def is_tar(path: str) -> bool:
+    """Return True if path has .tar and tarfile confirms tar archive"""
+    try:
+        with tarfile.open(path, 'r:*') as tar:
+            tar.getmembers()
+        return True
+    except (tarfile.TarError, EOFError, OSError, RuntimeError) as e:
+        # Not a valid tar file
+        print(f"Invalid tar file: {e}")
+        return False
+    print(f"Invalid tar file, other")
+    return False
+
+def is_within_directory(directory, target):
+    abs_dir = os.path.abspath(directory)
+    abs_target = os.path.abspath(target)
+    return abs_target.startswith(abs_dir)
+
+# Probably would be better to check for safety then pass to installed program, but seems like every compression has their own extractor
+class TarFile:
+    """Check for safe paths inside .tar and extraction"""
+
+    def __init__(self, path: str):
+        self.path = path
+        # Check if it's actually a tar-file
+        if not is_tar(self.path):
+            raise TypeError("File is not a .tar file")
+
+    def is_tar_safe(self, path="."):
+        try:
+            with tarfile.open(self.path, 'r:*') as tar:
+                for member in tar.getmembers():
+                    # Avoid files with absolute paths or .. (path traversal)
+                    if member.name.startswith("/") or ".." in member.name:
+                        raise Exception(f"Unsafe path detected: {member.name}")
+
+                    member_path = os.path.join(path, member.name)
+                    if not is_within_directory(path, member_path):
+                        raise Exception(f"Unsafe path detected: {member.name}")
+                    if member.issym() or member.islnk():
+                        raise Exception(f"Link detected: {member.name}")
+
+            return True
+        except tarfile.TarError:
+            # Not a valid tar file
+            return False
+
+
+    def namelist(self) -> List[str]:
+        """Return list of names in tar"""
+        names = []
+        try:
+            with tarfile.open(self.path, 'r:*') as tar:
+                for member in tar.getmembers():
+                    # Avoid files with absolute paths or .. (path traversal)
+                    if member.name.startswith("/") or ".." in member.name:
```

Comment:
Did you see the validations that Python itself implements for safe tar's?
https://github.com/python/cpython/blob/main/Lib/tarfile.py#L769
On a quick glance I can see they also check `os.sep` and they use `os.path.isabs` checks.
Maybe more to explore?